include/hve/arch/intel_x64/irq.h

```diff
@@ -0,0 +1,159 @@
+#include <utility>
+
+#include <bfgsl.h>
+#include <bfdebug.h>
+#include <bfvmm/hve/arch/intel_x64/vmcs/vmcs.h>
+#include <bfvmm/hve/arch/intel_x64/exit_handler/exit_handler.h>
+
+// -----------------------------------------------------------------------------
+// Exports
+// -----------------------------------------------------------------------------
+
+#include <bfexports.h>
+
+#ifndef STATIC_EAPIS_HVE
+#ifdef SHARED_EAPIS_HVE
+#define EXPORT_EAPIS_HVE EXPORT_SYM
+#else
+#define EXPORT_EAPIS_HVE IMPORT_SYM
+#endif
+#else
+#define EXPORT_EAPIS_HVE
+#endif
+
+#ifdef _MSC_VER
+#pragma warning(push)
+#pragma warning(disable : 4251)
+#endif
+
+namespace eapis
+{
+namespace intel_x64
+{
+
+namespace reason = ::intel_x64::vmcs::exit_reason::basic_exit_reason;
+
+// -----------------------------------------------------------------------------
+// Definitions
+// -----------------------------------------------------------------------------
+
+///
+/// External interrupt handler
+///
+class EXPORT_EAPIS_HVE irq
+{
+public:
+
+    struct info_t {
+        ::intel_x64::vmcs::value_type vec;
+    };
+
+    using vector_t = decltype(info_t::vec);
+    using irq_t = eapis::intel_x64::irq;
+    using vmcs_t = bfvmm::intel_x64::vmcs;
+    using exit_handler_t = bfvmm::intel_x64::exit_handler;
+    using handler_t = delegate<bool(gsl::not_null<vmcs_t *>, info_t &)>;
+
+    /// Constructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    irq(gsl::not_null<exit_handler_t *> exit_handler);
+
+    /// Destructor
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    ~irq() = default;
+
+    /// Add handler
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param vector the interrupt vector to listen to
+    /// @param d the handler to call when an exit occurs at vector
+    ///
+    void add_handler(vector_t vector, handler_t &&d);
+
+    /// Trap
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Enable external-interrupt exiting. This vcpu will exit each time
+    /// an interrupt arrives during VMX-nonroot operation.
+    ///
+    void trap();
+
+    /// Pass Through
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// Disable external-interrupt exiting. This vcpu will not exit if
+    /// the vcpu is interrupted during VMX-nonroot operation.
+    ///
+    void pass_through();
+
+    /// Handle
+    ///
+    /// Invoke the handler listening for irq exits at the
+    /// vector in the vmcs exit interruption info field
+    ///
+    /// @expects
+    /// @ensures
+    ///
+    /// @param vmcs the vmcs state passed to each handler
+    ///
+    bool handle(gsl::not_null<vmcs_t *> vmcs);
+
+    /// @cond
+
+    irq(irq &&) = default;
+    irq &operator=(irq &&) = default;
+
+    irq(const irq &) = delete;
+    irq &operator=(const irq &) = delete;
+
+    /// @endcond
+
+private:
+    exit_handler_t *m_exit_handler;
+    std::array<std::list<handler_t>, 256> m_handlers{};
+};
+
+} // namespace intel_x64
```

Comment:
we aren't labeling these in the other files, so for now I would prefer that we stick to that. Now... that being said, if we want to add them, I would merge a PR that did it consistently.